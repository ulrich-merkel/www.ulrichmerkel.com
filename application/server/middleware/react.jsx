/**
 * Es6 module for react server middleware.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires react-dom
 * @requires react-router
 * @requires lodash
 * @requires common/config/application
 * @requires common/config/routes
 * @requires common/utils/logger
 * @requires common/utils/read-file
 * @requires common/component/root
 * @requires common/component/layout/html
 * @requires common/component/page/not-found
 * @requires common/state/configure-store
 * @requires common/state/config/actions
 * @requires common/state/intl/actions
 * @requires common/state/csrf/actions
 *
 * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
 * @see {@link https://github.com/reactjs/react-router/issues/1990}
 * @see {@link https://www.youtube.com/watch?v=PnpfGy7q96U}
 * @see {@link https://github.com/reactjs/redux/issues/723}
 *
 * @changelog
 * - 0.0.3 Adjusted async rendering
 * - 0.0.2 Moved code to es6
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { RouterContext, match } from 'react-router';
import { get } from 'lodash';

import configApplication from './../../common/config/application';
import configRoutes from './../../common/config/routes';
import logger from './../../common/utils/logger';
import { readFile } from './../../common/utils/read-file';
import Root from './../../common/component/root';
import LayoutHtml from './../../common/component/layout/html';
import PageNotFound from './../../common/component/page/not-found';
import configureStore from './../../common/state/configure-store';
import { changeLocale } from './../../common/state/intl/actions';
import { fetchConfigContentIfNeeded, fetchConfigTranslationIfNeeded } from './../../common/state/config/actions';
import { addToken } from './../../common/state/csrf/actions';

const { aboveTheFold } = configApplication;

/**
 * Helper function to pass props to children components.
 *
 * @function
 * @private
 * @param {Object} Component The current router react component
 * @param {Object} props The current react component props
 * @returns {React.Element} React component markup
 */
function createElement(Component, props) {
    return <Component {...props} />;
}

/**
 * Using createElement function here to be able to
 * pass props to the page Component
 *
 * @see {@link https://github.com/reactjs/react-router/issues/3183}
 * @see {@link https://github.com/voronianski/universal-react-router-flux-2016/blob/master/src/server/middleware/renderHTML.js#L17}
 * @see {@link https://github.com/ReactTraining/react-router/issues/902}
 *
 * @function
 * @private
 * @param {Object} store The created redux store
 * @param {Object} renderProps The component properties to be rendered
 * @param {string} [cssBase=''] The file contents from base.css
 * @param {string} [scriptLoader=''] The file contents from loader.js
 * @returns {void}
 */
function getHtml(store, renderProps, cssBase = '', scriptLoader = '') {
    return renderToStaticMarkup(
        <Root store={store}>
            <LayoutHtml {... { store, cssBase, scriptLoader }}>
                <RouterContext
                    {...renderProps}
                    createElement={createElement}
                />
            </LayoutHtml>
        </Root>
    );
}

/**
 * Handle react server rendering and react routering.
 *
 * @function
 * @param {Object} req The current request object
 * @param {Object} res The result object
 * @returns {void}
 */
function middlewareReact(req, res, next) {

    /**
     * Using the new match function instead of Router.run
     *
     * Note that req.url here should be the full URL path from
     * the original request, including the query string.
     *
     * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
     * @see {@link https://github.com/reactjs/react-router/blob/master/docs/guides/ServerRendering.md}
     */
    match({
        routes: configRoutes,
        location: req.url
    }, (error, redirectLocation, renderProps) => {

        if (error) {
            return res.status(error.status || 500).send(error.message).end();
        }

        if (redirectLocation) {
            return res.redirect(res.status || 302, `${redirectLocation.pathname}${redirectLocation.search}`);
        }

        if (renderProps) {

            // Initialize redux store, combining single stores into a global one
            const store = configureStore();

            /**
             * Read language which is given in the url locale parameter or via
             * express middleware and store the selection in redux
             */
            const urlLocale = get(renderProps, 'router.params.locale', '');
            store.dispatch(changeLocale([urlLocale, urlLocale.toUpperCase()].join('-'), req.language));
            const acceptedLocale = get(store.getState(), 'intl.locale');

            /**
             * We check renderProps.components (or renderProps.routes) for
             * the "not found" component or route respectively, and send a 404 as
             * below, because we're using a catch-all error route in the router
             * config.
             *
             * @see {@link https://github.com/ReactTraining/react-router/issues/2834}
             */
            const statusCode = get(renderProps, 'components', []).indexOf(PageNotFound) !== -1 ? 404 : 200;

            /**
             * Load all required data async via promises to
             * improve the overall node performance.
             */
            return Promise.all([
                readFile(aboveTheFold.baseCss),
                readFile(aboveTheFold.loaderBundle),
                store.dispatch(fetchConfigContentIfNeeded()),
                store.dispatch(fetchConfigTranslationIfNeeded(acceptedLocale)),
                store.dispatch(addToken(req.csrfToken && req.csrfToken()))
            ]).then((result) => {
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps, result[0], result[1])}`);
            }).catch((reason) => {
                logger.warn(reason);
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps)}`);
            });

        }

        return next();

    });
}

export default middlewareReact;
