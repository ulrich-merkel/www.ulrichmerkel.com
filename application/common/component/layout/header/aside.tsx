/**
 * Es6 module for React Component.
 * Layout components merge modules to bigger parts of the
 * overall page layout.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, useCallback } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import classnames from 'classnames';
import { get, noop } from 'lodash';

import { url } from '../../../config/application';
import { addContent } from '../../decorator/add-content';
import { getContentSection } from '../../../utils/content';
import { Event, eventPreventDefault } from '../../../utils/event';
import {
    selectStateIntlLocale,
    selectStateIntlAvailableLocales
} from '../../../state/intl/selector';
import {
    changeIntlLocale,
    INTL_LOCALE_EN_EN,
    INTL_LOCALE_DE_DE
} from '../../../state/intl/duck';
import { Locale, Locales } from '../../../state/intl/types';
import {
    changeDialogVisibleSearch,
    changeDialogVisibleTheme
} from '../../../state/dialog/duck';
import { A } from '../../element/a';
import { Button } from '../../element/button';
import { Icon } from '../../element/icon';
import { List } from '../../element/list';
import { ModuleMenuListItem } from '../../module/menu/list-item';
import { isValidArray } from '../../../utils/array';

type Props = {
    className?: string;
    content?: {
        menu?: Record<string, unknown>;
    };
    intlAvailableLocales: Locales;
    intlLocale: Locale;
    onChangeDialogVisibleSearch?: (visible: boolean) => void;
    onChangeDialogVisibleTheme?: (visible: boolean) => void;
    onChangeIntlLocale?: (locale: string) => void;
};

/**
 * Get language button class names.
 *
 * @private
 * @param {string} intlLocale - The current selected locale
 * @param {string} buttonLocale - The buttons locale
 * @param {Array<string>} intlAvailableLocales - All available locales
 * @returns {string} The combined button class names
 */
function getIntlButtonClassNames(
    intlLocale: Locale,
    buttonLocale: Locale,
    intlAvailableLocales: Locales
): string {
    return classnames('m-menu__item', 'm-menu__item--en', {
        'is-active': intlLocale === buttonLocale,
        'is-hidden':
            !isValidArray(intlAvailableLocales) ||
            intlAvailableLocales.length < 2
    });
}

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} props - The current component props
 * @param {Function} props.handleIntlChangeLocale - Function handling language state changes
 * @param {Function} props.onChangeDialogVisibleSearch - Function handling dialog state changes
 * @param {Array.<string>} props.intlAvailableLocales - All available locale strings
 * @param {string} props.intlLocale - The current locale string
 * @param {string} [props.className] - The component css class names - will be merged into component default classNames
 * @param {object} [props.content={}] - The component content config
 * @returns {ReactElement} React component markup
 */
export const LayoutHeaderAside: FunctionComponent<Props> = (props) => {
    const {
        className,
        content,
        intlAvailableLocales,
        intlLocale,
        onChangeDialogVisibleSearch = noop,
        onChangeDialogVisibleTheme = noop,
        onChangeIntlLocale = noop
    } = props;

    const contentSection = getContentSection(content);
    const buttonEnClassName = getIntlButtonClassNames(
        intlLocale,
        INTL_LOCALE_EN_EN,
        intlAvailableLocales
    );
    const buttonDeClassName = getIntlButtonClassNames(
        intlLocale,
        INTL_LOCALE_DE_DE,
        intlAvailableLocales
    );
    const componentClassName = classnames(
        'm-menu--aside',
        'm-nav__toggle-target',
        'm-nav__aside',
        className
    );

    const handleChangeDialogVisibleSearch = useCallback(function fnCallback(
        event: Event
    ) {
        eventPreventDefault(event);
        onChangeDialogVisibleSearch(true);
    },
    []);

    const handleChangeDialogVisibleTheme = useCallback(function fnCallback(
        event: Event
    ) {
        eventPreventDefault(event);
        onChangeDialogVisibleTheme(true);
    },
    []);

    const handleIntlChangeLocale = useCallback(function fnCallback(
        event: Event
    ) {
        eventPreventDefault(event);
        const locale = get(event, 'target.dataset.locale');
        onChangeIntlLocale(locale);
    },
    []);

    return (
        <List
            className={componentClassName}
            itemType="http://schema.org/ItemList"
            role="menu"
        >
            <ModuleMenuListItem>
                <Button
                    className={buttonEnClassName}
                    dataLocale={INTL_LOCALE_EN_EN}
                    isClear
                    isSmall
                    onClick={handleIntlChangeLocale}
                    role="menuitem"
                    testId="button-language-en"
                    title={contentSection('menu.language.list[0].title')}
                >
                    {contentSection('menu.language.list[0].label')}
                </Button>
            </ModuleMenuListItem>
            <ModuleMenuListItem>
                <Button
                    className={buttonDeClassName}
                    dataLocale={INTL_LOCALE_DE_DE}
                    isClear
                    isSmall
                    onClick={handleIntlChangeLocale}
                    role="menuitem"
                    testId="button-language-de"
                    title={contentSection('menu.language.list[1].title')}
                >
                    {contentSection('menu.language.list[1].label')}
                </Button>
            </ModuleMenuListItem>
            <ModuleMenuListItem className="m-menu__list-item--search always-float">
                <A
                    className="m-menu__item--search c-btn--small c-btn--clear"
                    onClick={handleChangeDialogVisibleSearch}
                    role="menuitem"
                    testId="button-search"
                    title="Search"
                    to={url.search}
                >
                    <span className="c-btn__label">
                        <Icon className="c-btn__icon" icon="search" />
                    </span>
                </A>
            </ModuleMenuListItem>
            <ModuleMenuListItem className="m-menu__list-item--theme always-float">
                <A
                    className="m-menu__item--theme c-btn--small c-btn--clear"
                    onClick={handleChangeDialogVisibleTheme}
                    role="menuitem"
                    testId="button-settings"
                    title="Settings"
                    to={url.settings}
                >
                    <span className="c-btn__label">
                        <Icon className="c-btn__icon" icon="cog" />
                    </span>
                </A>
            </ModuleMenuListItem>
        </List>
    );
};

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the component’s props.
 *
 * @private
 * @param {object.<*>} state - The redux store state
 * @param {object.<*>} [ownProps] - The current component props
 * @returns {object}
 */
function mapStateToProps(state, ownProps) {
    return {
        intlAvailableLocales:
            selectStateIntlAvailableLocales(state) ||
            ownProps.intlAvailableLocales,
        intlLocale: selectStateIntlLocale(state) || ownProps.intlLocale
    };
}

/**
 * If an object is passed, each function inside it will be assumed to
 * be a Redux action creator. An object with the same function names,
 * but with every action creator wrapped into a dispatch call so they
 * may be invoked directly, will be merged into the component’s props.
 * If a function is passed, it will be given dispatch.
 *
 * @private
 * @type {object<string, Function>}
 */
const mapDispatchToProps = {
    onChangeIntlLocale: changeIntlLocale,
    onChangeDialogVisibleSearch: changeDialogVisibleSearch,
    onChangeDialogVisibleTheme: changeDialogVisibleTheme
};

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {ReactElement}
 */
export const LayoutHeaderAsideConnected = withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )(addContent('LayoutHeaderAside')(LayoutHeaderAside))
);
