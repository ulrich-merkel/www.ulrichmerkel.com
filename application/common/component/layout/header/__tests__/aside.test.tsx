import * as React from 'react';
import {
    render,
    screen,
    userEvent
} from '../../../../../__tests__/utils/test-utils';
import { LayoutHeaderAside, LayoutHeaderAsideConnected } from '../aside';

describe('LayoutHeaderAside', function fnDescribe() {
    const props = {
        className: 'layout-header',
        content: {},
        intlAvailableLocales: ['en-EN', 'de-DE'],
        intlLocale: 'de-DE',
        onChangeDialogVisibleSearch: jest.fn(),
        onChangeDialogVisibleTheme: jest.fn(),
        onChangeIntlLocale: jest.fn()
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <LayoutHeaderAside {...props}>Header Children</LayoutHeaderAside>
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly with default props', function fnIt() {
        const { asFragment } = render(<LayoutHeaderAside />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly the connected version', function fnIt() {
        const { asFragment } = render(
            <LayoutHeaderAsideConnected {...props}>
                Header Children
            </LayoutHeaderAsideConnected>
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it('should show search dialog', function fnIt() {
        const onChangeDialogVisibleSearch = jest.fn();
        render(
            <LayoutHeaderAside
                {...props}
                {...{ onChangeDialogVisibleSearch }}
            />
        );

        const button = screen.getByTestId('button-search');
        userEvent.click(button);
        expect(onChangeDialogVisibleSearch).toHaveBeenCalled();
    });
    it('should show settings dialog', function fnIt() {
        const onChangeDialogVisibleTheme = jest.fn();
        render(
            <LayoutHeaderAside {...props} {...{ onChangeDialogVisibleTheme }} />
        );

        const button = screen.getByTestId('button-settings');
        userEvent.click(button);
        expect(onChangeDialogVisibleTheme).toHaveBeenCalled();
    });
    it('should change languages', function fnIt() {
        const onChangeIntlLocale = jest.fn();
        render(<LayoutHeaderAside {...props} {...{ onChangeIntlLocale }} />);

        const button = screen.getByTestId('button-language-de');
        userEvent.click(button);
        expect(onChangeIntlLocale).toHaveBeenCalled();
    });
});
