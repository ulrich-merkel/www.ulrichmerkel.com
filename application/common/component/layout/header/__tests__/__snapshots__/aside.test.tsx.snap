// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutHeaderAside should render correctly 1`] = `
<DocumentFragment>
  <ul
    class="m-menu--aside m-nav__toggle-target m-nav__aside layout-header"
    itemscope=""
    itemtype="http://schema.org/ItemList"
    role="menu"
  >
    <li
      class="m-menu__list-item"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <button
        class="c-btn m-menu__item m-menu__item--en c-btn--small c-btn--clear"
        data-locale="en-EN"
        role="button"
        title=""
        type="button"
      >
        <span
          class="c-btn__label"
        />
      </button>
    </li>
    <li
      class="m-menu__list-item"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <button
        class="c-btn m-menu__item m-menu__item--de is-active c-btn--small c-btn--clear"
        data-locale="de-DE"
        role="button"
        title=""
        type="button"
      >
        <span
          class="c-btn__label"
        />
      </button>
    </li>
    <li
      class="m-menu__list-item--search always-float"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <a
        class="c-link m-menu__item--search c-btn--small c-btn--clear"
        href="/search"
        tabindex="0"
        title="Search"
      >
        <span
          class="c-btn__label"
        >
          <i
            aria-hidden="true"
            class="c-font-icon--search c-btn__icon"
          />
        </span>
      </a>
    </li>
    <li
      class="m-menu__list-item--theme always-float"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <a
        class="c-link m-menu__item--theme c-btn--small c-btn--clear"
        href="/settings"
        tabindex="0"
        title=""
      >
        <span
          class="c-btn__label"
        >
          <i
            aria-hidden="true"
            class="c-font-icon--cog c-btn__icon"
          />
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`LayoutHeaderAside should render correctly the connected version 1`] = `
<DocumentFragment>
  <ul
    class="m-menu--aside m-nav__toggle-target m-nav__aside layout-header"
    itemscope=""
    itemtype="http://schema.org/ItemList"
    role="menu"
  >
    <li
      class="m-menu__list-item"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <button
        class="c-btn m-menu__item m-menu__item--en is-active is-hidden c-btn--small c-btn--clear"
        data-locale="en-EN"
        role="button"
        title=""
        type="button"
      >
        <span
          class="c-btn__label"
        />
      </button>
    </li>
    <li
      class="m-menu__list-item"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <button
        class="c-btn m-menu__item m-menu__item--de is-hidden c-btn--small c-btn--clear"
        data-locale="de-DE"
        role="button"
        title=""
        type="button"
      >
        <span
          class="c-btn__label"
        />
      </button>
    </li>
    <li
      class="m-menu__list-item--search always-float"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <a
        class="c-link m-menu__item--search c-btn--small c-btn--clear"
        href="/search"
        tabindex="0"
        title="Search"
      >
        <span
          class="c-btn__label"
        >
          <i
            aria-hidden="true"
            class="c-font-icon--search c-btn__icon"
          />
        </span>
      </a>
    </li>
    <li
      class="m-menu__list-item--theme always-float"
      itemprop="itemListElement"
      itemscope=""
      itemtype="http://www.schema.org/SiteNavigationElement"
    >
      <a
        class="c-link m-menu__item--theme c-btn--small c-btn--clear"
        href="/settings"
        tabindex="0"
        title=""
      >
        <span
          class="c-btn__label"
        >
          <i
            aria-hidden="true"
            class="c-font-icon--cog c-btn__icon"
          />
        </span>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
